File: LeetCode root folder 5/Partition Array to Minimize XOR.py
# no other way, even this simple DP, my last resort, failed
class Solution:
    def minXor(self, nums: List[int], k: int) -> int:
        inf = math.inf
        n = len(nums)
        dp = nums[:]

        for i in range(n - 1, -1, -1):
            xor = 0
            for j in range(i, n):
                xor ^= nums[j]
                dp[i] = min(dp[i], max(xor, dp[j + 1] if j < n - 1 else 0))
        return dp[0]

# TLE
class Solution:
    def minXor(self, nums: List[int], k: int) -> int:
        inf = math.inf
        n = len(nums)
        arr = [[-1]*k for _ in range(n)]
        def recursive(i, k):
            nonlocal inf, n
            if k < 0: return inf
            if i >= n: return 0 if k == 0 else inf
            if arr[i][k-1] != -1: return arr[i][k-1]
            res = inf
            xor = 0
            for j in range(i, n):
                xor ^= nums[j]
                a = recursive(j+1, k-1)
                res = min(res, max(a, xor))
            arr[i][k-1] = res
            return res
        return recursive(0, k)

File: LeetCode root folder 5/Partition Array to Minimize XOR.cpp
// hate leetcode when it being super unfriendly to python user, and I can't believe I forget to implement DP during
// the contest and wondering why this hit TLE sooner than my python code...: 43%
class Solution {
public:
    vector<int> arr;
    int n = -1;
    int **dp;
    long recursive(int i, int k) {
        if (i >= n) {
            if (k == 0) return 0;
            else return 1<<30;
        }
        if (k < 1) return 1<<30;
        if (dp[i][k-1] != -1) return dp[i][k-1];
        long res = 1<<30;
        long x = 0;
        for (int j = i; j<n; j++) {
            x ^= arr[j];
            long a = recursive(j+1, k-1);
            res = min(res, max(a, x));
        }
        dp[i][k-1] = res;
        return res;
    }

    long minXor(vector<int>& nums, int k) {
        n = nums.size();
        arr = nums;
        dp = (int **)malloc(sizeof(int *)*n);
        for (int i = 0; i < n; i++) {
            dp[i] = (int *)malloc(sizeof(int)*k);
            memset(dp[i], -1, sizeof(int)*k);
        }
        int res = recursive(0, k);
        for (int i = 0; i < n; i++) free(dp[i]);
        free(dp);
        return res;
    }
};